src/mlpack/core/data/load_impl.hpp

```diff
@@ -20,376 +21,300 @@
+  bool success = OpenFile(filename, opts, true, stream);
+  if (!success)
   {
     Timer::Stop("loading_data");
-    if (fatal)
-      Log::Fatal << "Cannot open file '" << filename << "'. " << std::endl;
-    else
-      Log::Warn << "Cannot open file '" << filename << "'; load failed."
-          << std::endl;
-
     return false;
   }
 
-  if (extension == "csv" || extension == "tsv" || extension == "txt")
+  success = DetectFileType<MatType>(filename, opts, true, &stream);
+  if (!success)
   {
-    Log::Info << "Loading '" << filename << "' as CSV dataset.  " << std::flush;
-    try
-    {
-      LoadCSV loader(filename);
-      loader.LoadCategoricalCSV(matrix, info, transpose);
-    }
-    catch (std::exception& e)
-    {
-      Timer::Stop("loading_data");
-      if (fatal)
-        Log::Fatal << e.what() << std::endl;
-      else
-        Log::Warn << e.what() << std::endl;
+    Timer::Stop("loading_data");
+    return false;
+  }
 
-      return false;
-    }
+  if constexpr (IsArma<MatType>::value || IsSparseMat<MatType>::value)
+  {
+    // compile time type detction
+    // convert this to a move, and move back at the end.
+    // add a move constructor that does this.
+    TextOptions txtOpts(std::move(opts));
+    success = LoadMatrix(filename, matrix, stream, txtOpts);
+    opts = std::move(txtOpts);
   }
-  else if (extension == "arff")
+  else 
   {
-    Log::Info << "Loading '" << filename << "' as ARFF dataset.  "
-        << std::flush;
-    try
-    {
-      LoadARFF(filename, matrix, info);
-
-      // We transpose by default.  So, un-transpose if necessary...
-      if (!transpose)
-      {
-        return inplace_transpose(matrix, fatal);
-      }
-    }
-    catch (std::exception& e)
-    {
-      Timer::Stop("loading_data");
-      if (fatal)
-        Log::Fatal << e.what() << std::endl;
-      else
-        Log::Warn << e.what() << std::endl;
-
-      return false;
-    }
+    throw std::runtime_error("DataOptionType is unknown!."
+        "please use a known type or provide specific overloads");
   }
-  else
+
+  if (!success)
   {
-    // The type is unknown.
+    Log::Info << std::endl;
     Timer::Stop("loading_data");
-    if (fatal)
-      Log::Fatal << "Unable to detect type of '" << filename << "'; "
-          << "incorrect extension?" << std::endl;
+    if (opts.Fatal())
+      Log::Fatal << "Loading from '" << filename << "' failed." << std::endl;
     else
-      Log::Warn << "Unable to detect type of '" << filename << "'; load failed."
-          << " Incorrect extension?" << std::endl;
+      Log::Warn << "Loading from '" << filename << "' failed." << std::endl;
 
     return false;
   }
-
-  Log::Info << "Size is " << (transpose ? matrix.n_cols : matrix.n_rows)
-      << " x " << (transpose ? matrix.n_rows : matrix.n_cols) << ".\n";
+  else
+  {
+   if constexpr (IsArma<MatType>::value)
+   {
+      Log::Info << "Size is " << matrix.n_rows << " x "
+                << matrix.n_cols << ".\n";
+   }
```

Comment:
```suggestion
    if constexpr (IsArma<MatType>::value)
    {
       Log::Info << "Size is " << matrix.n_rows << " x "
                 << matrix.n_cols << ".\n";
    }
```

I think we could remove the `IsArma` check---any matrix type is already guaranteed to be an Armadillo-like type and have `.n_rows` and `.n_cols`.

Reply:
This is necessary when you have a model that you are loading or saving, the model does not have the n_cols or n_rows. Let us keep it for now so we do not forget adding it later